feat(archives): add size field to archived recordings

lkonno opened the pull request:
Fixes [#964](https://github.com/cryostatio/cryostat/issues/964)

- Added size attribute to `ArchivedRecordingInfo`;

`$ https :8181/api/v1/recordings`

- Added size to graphql query result

`http :8181/api/v2.2/graphql query="query { targetNodes { name nodeType labels recordings { archived { data { name metadata {labels} size } aggregate {count size} } } } }"`

`http :8181/api/v2.2/graphql query="query { archivedRecordings { name downloadUrl reportUrl metadata { labels } size } }"`

- Added the sizeBytesGreaterThanEqual and sizeBytesLessThanEqual filters to both above;

`http :8181/api/v2.2/graphql query="query { targetNodes { name nodeType labels recordings { archived (filter:{sizeBytesGreaterThanEqual:2578134}) { data { name metadata {labels} size } aggregate {count size} } } } }"`

`http :8181/api/v2.2/graphql query="query { archivedRecordings (filter:{sizeBytesGreaterThanEqual:2578134}) { name downloadUrl reportUrl metadata { labels } size } }"`

andrewazores commented:
```
$ http :8181/api/v2.2/graphql Authorization:"Basic $(echo user:pass | base64)" query="{ archivedRecordings { aggregate { size count } } }"
HTTP/1.1 200 OK
content-encoding: gzip
content-length: 83
content-type: application/json
vary: origin

{
    "data": {
        "archivedRecordings": {
            "aggregate": {
                "count": 0,
                "size": 0
            }
        }
    }
}
```

```
$ http :8181/api/v2.2/graphql Authorization:"Basic $(echo user:pass | base64)" query="query { targetNodes { name recordings { archived { aggregate { count } } }
 } }"
HTTP/1.1 200 OK
content-encoding: gzip
content-length: 137
content-type: application/json
vary: origin

{
    "data": {
        "targetNodes": [
            {
                "name": "service:jmx:rmi:///jndi/rmi://carbon:9091/jmxrmi",
                "recordings": {
                    "archived": {
                        "aggregate": {
                            "count": 0
                        }
                    }
                }
            }
        ]
    }
}
```

Looks like both queries on archived recordings are working. I made one tiny change locally:

```diff
diff --git a/src/main/resources/queries.graphqls b/src/main/resources/queries.graphqls
index 176e24ad..1a90bc0f 100644
--- a/src/main/resources/queries.graphqls
+++ b/src/main/resources/queries.graphqls
@@ -2,5 +2,5 @@ type Query {
     rootNode: EnvironmentNode!
     environmentNodes(filter: EnvironmentNodeFilterInput): [EnvironmentNode!]!
     targetNodes(filter: TargetNodesFilterInput): [TargetNode!]!
-    archivedRecordings(filter: ArchivedRecordingFilterInput): [Archived!]!
+    archivedRecordings(filter: ArchivedRecordingFilterInput): Archived!
 }
```

:-)

I briefly reviewed the implementation and it's looking good so far. LMK when you're ready for full review.

lkonno commented:
Thank you for finding the issue! 
I believe it is ready for review now.

andrewazores commented:
Rebase the branch and resolve the merge conflict listed above, please.

andrewazores commented on src/main/java/io/cryostat/rules/ArchivedRecordingInfo.java:
```diff
@@ -67,6 +68,22 @@ public ArchivedRecordingInfo(
         this.downloadUrl = downloadUrl;
         this.reportUrl = reportUrl;
         this.metadata = metadata;
+        this.size = 0;
```
This constructor is also used in a unit test, so similar comment as the previous.

It's also used in `RecordingsPostHandler` which is the endpoint for re-uploading JFR files into Cryostat's archives (related: https://github.com/cryostatio/cryostat/pull/1047) and `PutArchivedRecordingMetadataMutator`, which is the GraphQL mutator for updating metadata/labels on archived recordings.

When a file is uploaded to Cryostat's archives, we can easily check the size of that file and use a real value here rather than `0`.

In the case of updating metadata related to an archived recording, the mutator actually already has an `ArchivedRecordingInfo` context object to begin with. These are immutable and so the mutator returns a new object with copied values and an updated/replaced `Metadata`. The size field can also just be copied from the context object.

lkonno replied:
done!
In the tests, I just passed 0 value to file size.

andrewazores reviewed (commented):
I think some of the other example GraphQL queries in https://github.com/cryostatio/cryostat/tree/main/docs/graphql will also need to be updated to match the new `data / aggregate` structure

andrewazores commented on src/main/java/io/cryostat/net/web/http/api/v2/graph/ArchivedRecordingsFetcher.java:
```diff
@@ -113,5 +128,6 @@ static class Archived {
 
     static class AggregateInfo {
         Long count;
+        Long size;
```
Not sure why I didn't catch this when the `count` was added, but can these two fields be changed to the primitive `long` instead?

lkonno replied:
done!

lkonno commented on src/main/java/io/cryostat/rules/ArchivedRecordingInfo.java:
```diff
@@ -50,10 +50,11 @@ public class ArchivedRecordingInfo {
     public ArchivedRecordingInfo(
             String encodedServiceUri, String name, String downloadUrl, String reportUrl) {
-        this(encodedServiceUri, name, downloadUrl, reportUrl, new Metadata());
+        this(encodedServiceUri, name, downloadUrl, reportUrl, new Metadata(), 0);
```
done!
In the tests, I just added empty `new Metadata()`.

andrewazores reviewed (commented):
Looks great other than that one small adjustment to a sample query!

lkonno commented on docs/graphql/environment-nodes-query-3.graphql:
```diff
@@ -16,6 +16,11 @@ query {
+                    aggregate {
+                        count
+                        size
+                    }
```
done!

andrewazores commented:
Sorry, one last thing I just thought of. Could you update the `GraphQLIT` integration test to also verify the new size field?

lkonno commented:
I updated the archived classes in `GraphQLIT` and updated the tests containing archived recording to check the size to not be null. Is it fine or do you think that it could be a new test?

andrewazores commented on src/test/java/itest/GraphQLIT.java:
```diff
@@ -483,6 +484,7 @@ void testDeleteMutation() throws Exception {
         MatcherAssert.assertThat(node.recordings.active, Matchers.hasSize(1));
         MatcherAssert.assertThat(node.recordings.archived.data, Matchers.hasSize(1));
         MatcherAssert.assertThat(node.recordings.archived.aggregate.count, Matchers.equalTo(1L));
+        MatcherAssert.assertThat(node.recordings.archived.aggregate.size, Matchers.notNullValue());
```
Could this use `Matchers.greaterThan(0L)` perhaps?

andrewazores commented on src/test/java/itest/GraphQLIT.java:
```diff
@@ -438,6 +438,7 @@ void testArchivedRecordingMetadataMutation() throws Exception {
         MatcherAssert.assertThat(node.recordings.archived.data, Matchers.hasSize(1));
 
         ArchivedRecording archivedRecording = node.recordings.archived.data.get(0);
+        MatcherAssert.assertThat(archivedRecording.size, Matchers.notNullValue());
```
Same here

lkonno replied:
the test failed, will fix it.

andrewazores commented:
I think it's fine as it is, no need to split it into a separate test.

lkonno commented:
I forgot to pass the size to query in the last test. I think it is fine now.
